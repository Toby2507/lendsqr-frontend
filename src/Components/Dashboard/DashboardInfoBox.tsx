interface infoboxInterface {
  img: string;
  title: string;
  amount: string;
}

const DashboardInfoBox = ({ img, title, amount }: infoboxInterface) => {
  return (
    <article className="dashboard-infobox__info">
      <figure className="dashboard-infobox__info--img">
        <img src={img} alt={title} />
      </figure>
      <h2 className="dashboard-infobox__info--title">{title}</h2>
      <span className="dashboard-infobox__info--amount">{amount}</span>
    </article>
  );
};

export default DashboardInfoBox;