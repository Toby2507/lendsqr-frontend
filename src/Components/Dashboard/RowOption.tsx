import React, { useState } from 'react';
import { HiDotsVertical } from 'react-icons/hi';
import { BiUserCheck, BiUserX } from 'react-icons/bi';
import { BsEye } from 'react-icons/bs';
import { Link } from 'react-router-dom';

const RowOption = ({ userId }: { userId: string; }) => {
  const [showOptions, setShowOptions] = useState<boolean>(false);
  return (
    <div className="row-options">
      <button className="row-options__btn" onClick={() => setShowOptions(!showOptions)}>
        <HiDotsVertical />
      </button>
      {showOptions && (
        <div className="row-options__box">
          <Link to={`/user/${userId}`}>
            <div className="row-options__box--item">
              <BsEye />
              <p>view details</p>
            </div>
          </Link>
          <div className="row-options__box--item">
            <BiUserX />
            <p>blacklist user</p>
          </div>
          <div className="row-options__box--item">
            <BiUserCheck />
            <p>activate user</p>
          </div>
        </div>
      )}
    </div>
  );
};

export default RowOption;