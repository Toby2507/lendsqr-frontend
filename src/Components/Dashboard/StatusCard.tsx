const StatusCard = ({ type }: { type: string; }) => {
  const values = {
    Blacklisted: { name: "blacklisted", className: "statusCard__red" },
    Inactive: { name: "inactive", className: "statusCard__blue" },
    Pending: { name: "pending", className: "statusCard__yellow" },
    ActiveUser: { name: "active", className: "statusCard__green" },
  };
  return (
    <span className={values[type as keyof typeof values].className}>{values[type as keyof typeof values].name}</span>
  );
};

export default StatusCard;