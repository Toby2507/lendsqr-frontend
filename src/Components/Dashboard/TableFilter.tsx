import { Column } from '@tanstack/react-table';
import { memo, useEffect, useMemo, useState } from 'react';
import { DebouncedInput } from './DebouncedInput';
import filterData from '../../Data/tableFilterOptions.json';

interface filterBoxInterface {
  column: Column<any, unknown>;
  filter: boolean;
  reset: boolean;
  setReset: React.Dispatch<React.SetStateAction<boolean>>;
  setFilter: React.Dispatch<React.SetStateAction<boolean>>;
  setShowBox: React.Dispatch<React.SetStateAction<boolean>>;
}

const TableFilter = ({ column, filter, reset, setFilter, setReset, setShowBox }: filterBoxInterface) => {
  const filterKey = column.id as keyof typeof filters;
  const columnFilterValue = column.getFilterValue();
  const [value, setValue] = useState<string>((columnFilterValue ?? '') as string);
  const filters = filterData;
  const sortedUniqueValues = useMemo(() => Array.from(column.getFacetedUniqueValues().keys()).sort(), [column]);

  useEffect(() => {
    if (filter) {
      column.setFilterValue(value);
      setTimeout(() => setShowBox(false), 100);
      setTimeout(() => setFilter(false), 200);
    }
    if (reset) {
      setValue("");
      setTimeout(() => column.setFilterValue(value), 100);
      setTimeout(() => setShowBox(false), 200);
      setTimeout(() => setReset(false), 300);
    }
  }, [reset, filter, column, value, setFilter, setReset, setShowBox]);

  return (
    <>
      <datalist id={column.id + 'list'}>
        {sortedUniqueValues.map((value, i) => <option value={value} key={i} />)}
      </datalist>
      <div className="table-filter__input">
        <label htmlFor={`search_${column.id}`} className="table-filter__input--label">{filters[filterKey].title}</label>
        <DebouncedInput
          type={filters[filterKey].type}
          id={`search_${column.id}`}
          value={value}
          onChange={value => setValue(value as string)}
          placeholder={filters[filterKey].placeholder}
          className="table-filter__input--input"
          list={column.id + 'list'}
        />
      </div>
    </>
  );
};

export default memo(TableFilter);