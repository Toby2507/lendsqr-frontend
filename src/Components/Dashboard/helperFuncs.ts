import { RankingInfo, rankItem } from '@tanstack/match-sorter-utils';
import { FilterFn, Row } from '@tanstack/react-table';

declare module '@tanstack/table-core' {
  interface FilterFns {
    fuzzy: FilterFn<unknown>;
  }
  interface FilterMeta {
    itemRank: RankingInfo;
  }
}

export const fuzzyFilter: FilterFn<any> = (row, columnId, value, addMeta) => {
  const itemRank = rankItem(row.getValue(columnId), value);
  addMeta({ itemRank });
  return itemRank.passed;
};

const contains = <TData extends Record<string, any> = {}>(
  row: Row<TData>,
  id: string,
  filterValue: string | number
) =>
  row
    .getValue<string | number>(id)
    .toString()
    .toLowerCase()
    .trim()
    .includes(filterValue.toString().toLowerCase().trim());

contains.autoRemove = (val: any) => !val;

const startsWith = <TData extends Record<string, any> = {}>(
  row: Row<TData>,
  id: string,
  filterValue: string | number
) =>
  row
    .getValue<string | number>(id)
    .toString()
    .toLowerCase()
    .trim()
    .startsWith(filterValue.toString().toLowerCase().trim());

startsWith.autoRemove = (val: any) => !val;