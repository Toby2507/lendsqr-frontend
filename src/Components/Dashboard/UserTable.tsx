import { ColumnDef, ColumnFiltersState, flexRender, getCoreRowModel, getFacetedRowModel, getFacetedUniqueValues, getFilteredRowModel, getPaginationRowModel, getSortedRowModel, useReactTable } from "@tanstack/react-table";
import { memo, useEffect, useState } from "react";
import { FaAngleLeft, FaAngleRight } from "react-icons/fa";
import { IoFilter } from "react-icons/io5";
import { userInterface } from "../../Utils/interfaces";
import TableFilter from "./TableFilter";
import { fuzzyFilter } from "./helperFuncs";

interface tableInterface {
  columns: ColumnDef<userInterface>[];
  data: userInterface[];
}
const UserTable = ({ columns, data }: tableInterface) => {
  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>([]);
  const [filterTable, setFilterTable] = useState<boolean>(false);
  const [resetTable, setResetTable] = useState<boolean>(false);
  const [showFilterBox, setShowFilterBox] = useState<boolean>(false);
  const filterBoxProps = {
    reset: resetTable,
    filter: filterTable,
    setReset: setResetTable,
    setFilter: setFilterTable,
    setShowBox: setShowFilterBox
  };
  const table = useReactTable({
    data,
    columns,
    filterFns: { fuzzy: fuzzyFilter },
    state: { columnFilters },
    onColumnFiltersChange: setColumnFilters,
    getCoreRowModel: getCoreRowModel(),
    getFilteredRowModel: getFilteredRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getPaginationRowModel: getPaginationRowModel(),
    getFacetedRowModel: getFacetedRowModel(),
    getFacetedUniqueValues: getFacetedUniqueValues()
  });

  const filterColumns = table.getHeaderGroups().map(headerGroup => (
    headerGroup.headers.filter(header => header.column.getCanFilter()).map(header => header.column)
  ));
  const currentPage = (table.getState().pagination.pageIndex + 1);
  const pages = Array.from({ length: table.getPageCount() + 1 - currentPage }, (v, i) => i + currentPage);

  useEffect(() => {
    if (table.getState().columnFilters[0]?.id === 'fullName') {
      if (table.getState().sorting[0]?.id !== 'fullName') {
        table.setSorting([{ id: 'fullName', desc: false }]);
      }
    }
  }, [table]);
  return (
    <div className="dashboard-wrapper">
      <section className="dashboard-table">
        {/* FILTER BOX */}
        {showFilterBox && (
          <div className="table-filter">
            {filterColumns[0].map((column, i) => <TableFilter key={i} column={column} {...filterBoxProps} />)}
            <div className="table-filter__action">
              <button className="table-filter__action--reset" onClick={() => setResetTable(true)}>reset</button>
              <button className="table-filter__action--filter" onClick={() => setFilterTable(true)}>filter</button>
            </div>
          </div>
        )}
        {/* DATA TABLE */}
        <table>
          <thead className="dashboard-table__header">
            {table.getHeaderGroups().map(headerGroup => (
              <tr key={headerGroup.id}>
                {headerGroup.headers.map(header => (
                  <th key={header.id}>
                    <div className="dashboard-table__header--heading" onClick={() => setShowFilterBox(!showFilterBox)}>
                      {header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())}
                      {!header.isPlaceholder && < IoFilter />}
                    </div>
                  </th>
                ))}
              </tr>
            ))}
          </thead>
          <tbody className="dashboard-table__body">
            {table.getRowModel().rows.map((row) => (
              <tr key={row.id}>
                {row.getVisibleCells().map((cell) => (
                  <td key={cell.id}>
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </section>
      {/* NAVIGATION */}
      <div className="dashboard-navigation">
        {/* ALTER PAGE SIZE */}
        <div className="dashboard-navigation__size">
          <p>Showing</p>
          <select
            value={table.getState().pagination.pageSize}
            onChange={e => { table.setPageSize(Number(e.target.value)); }}
          >
            {[10, 20, 25, 50].map(pageSize => <option key={pageSize} value={pageSize}>{pageSize}</option>)}
          </select>
          <p>out of {data.length}</p>
        </div>
        {/* NAVIGATE PAGES */}
        <div className="dashboard-navigation__page">
          <button
            className="dashboard-navigation__page--btn"
            onClick={() => table.previousPage()}
            disabled={!table.getCanPreviousPage()}
          ><FaAngleLeft /></button>
          {
            pages.map((page, idx) => {
              if (pages.length <= 5) return <p key={idx}>{page}</p>;
              if ([0, 1, 2].includes(idx)) return <p key={idx}>{page}</p>;
              if (idx === 4) return <p key={idx}>...</p>;
              if ([pages.length - 1, pages.length - 2].includes(idx)) return <p key={idx}>{page}</p>;
              return null;
            })
          }
          <button
            className="dashboard-navigation__page--btn"
            onClick={() => table.nextPage()}
            disabled={!table.getCanNextPage()}
          ><FaAngleRight /></button>
        </div>
      </div>
    </div>
  );
};

export default memo(UserTable);