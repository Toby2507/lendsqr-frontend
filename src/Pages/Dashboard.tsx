import { ColumnDef } from '@tanstack/react-table';
import dayjs from 'dayjs';
import { useMemo } from 'react';
import DashboardInfoBox from '../Components/Dashboard/DashboardInfoBox';
import StatusCard from '../Components/Dashboard/StatusCard';
import UserTable from '../Components/Dashboard/UserTable';
import data from '../Data/infoBoxData.json';
import { userInterface } from '../Utils/interfaces';
import { getFromLocal } from '../Utils/localStorage';
import RowOption from '../Components/Dashboard/RowOption';

const Dashboard = () => {
  const infoBoxData = data;
  const users: userInterface[] = getFromLocal("LendsqrUsers");
  const statuses = useMemo(() => ["Blacklisted", "Inactive", "Pending", "ActiveUser"], []);
  const columns = useMemo<ColumnDef<userInterface>[]>(() => [
    {
      header: 'organization',
      accessorKey: 'orgName',
      cell: (row) => row.renderValue()
    },
    {
      header: 'username',
      accessorKey: 'userName',
      cell: (row) => row.renderValue()
    },
    {
      header: 'email',
      accessorKey: 'email',
      cell: (row) => row.renderValue()
    },
    {
      header: 'phone number',
      accessorKey: 'phoneNumber',
      cell: (row) => row.renderValue()
    },
    {
      header: 'date joined',
      accessorFn: (row) => dayjs(row.createdAt).format('MMM DD, YYYY hh:mm A'),
      cell: (row) => row.renderValue()
    },
    {
      header: 'status',
      accessorFn: (row) => statuses[row.userStatus],
      cell: (row) => <StatusCard type={row.renderValue() as string} />
    },
    {
      header: ({ header }) => header.isPlaceholder = true,
      accessorKey: 'id',
      cell: (row) => <RowOption userId={row.renderValue() as string} />
    }
  ], [statuses]);

  return (
    <section className="dashboard">
      <h1 className="dashboard-title">Users</h1>
      <div className="dashboard-infobox">
        {infoBoxData.map((box, i) => <DashboardInfoBox key={i} {...box} />)}
      </div>
      <UserTable columns={columns} data={users} />
    </section>
  );
};

export default Dashboard;