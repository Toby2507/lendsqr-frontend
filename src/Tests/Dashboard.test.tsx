import Adapter from '@cfaester/enzyme-adapter-react-18';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dashboard from "../Pages/Dashboard";

Enzyme.configure({ adapter: new Adapter() });
describe("Dashboard Component Test", () => {
  test("should render without error", () => {
    const dashboard = shallow(<Dashboard />);
    expect(toJson(dashboard)).toMatchSnapshot();
  });
});