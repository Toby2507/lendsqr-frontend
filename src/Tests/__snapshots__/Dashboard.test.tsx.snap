// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Component Test should render without error 1`] = `
<section
  className="dashboard"
>
  <h1
    className="dashboard-title"
  >
    Users
  </h1>
  <div
    className="dashboard-infobox"
  >
    <DashboardInfoBox
      amount="2,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280917/all_users_bczc3b.svg"
      key="0"
      title="users"
    />
    <DashboardInfoBox
      amount="2,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280918/active_users_ptnp9q.svg"
      key="1"
      title="active users"
    />
    <DashboardInfoBox
      amount="12,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280917/user_loans_bu8eul.svg"
      key="2"
      title="users with loans"
    />
    <DashboardInfoBox
      amount="102,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280917/user_savings_aryv4v.svg"
      key="3"
      title="users with savings"
    />
  </div>
  <Memo(UserTable)
    columns={
      Array [
        Object {
          "accessorKey": "orgName",
          "cell": [Function],
          "header": "organization",
        },
        Object {
          "accessorKey": "userName",
          "cell": [Function],
          "header": "username",
        },
        Object {
          "accessorKey": "email",
          "cell": [Function],
          "header": "email",
        },
        Object {
          "accessorKey": "phoneNumber",
          "cell": [Function],
          "header": "phone number",
        },
        Object {
          "accessorFn": [Function],
          "cell": [Function],
          "header": "date joined",
        },
        Object {
          "accessorFn": [Function],
          "cell": [Function],
          "header": "status",
        },
        Object {
          "accessorKey": "id",
          "cell": [Function],
          "header": [Function],
        },
      ]
    }
    data={null}
  />
</section>
`;

exports[`Dashboard should render without error 1`] = `
<section
  className="dashboard"
>
  <h1
    className="dashboard-title"
  >
    Users
  </h1>
  <div
    className="dashboard-infobox"
  >
    <DashboardInfoBox
      amount="2,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280917/all_users_bczc3b.svg"
      key="0"
      title="users"
    />
    <DashboardInfoBox
      amount="2,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280918/active_users_ptnp9q.svg"
      key="1"
      title="active users"
    />
    <DashboardInfoBox
      amount="12,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280917/user_loans_bu8eul.svg"
      key="2"
      title="users with loans"
    />
    <DashboardInfoBox
      amount="102,453"
      img="https://res.cloudinary.com/db1nlq5lv/image/upload/v1675280917/user_savings_aryv4v.svg"
      key="3"
      title="users with savings"
    />
  </div>
  <Memo(UserTable)
    columns={
      Array [
        Object {
          "accessorKey": "orgName",
          "cell": [Function],
          "header": "organization",
        },
        Object {
          "accessorKey": "userName",
          "cell": [Function],
          "header": "username",
        },
        Object {
          "accessorKey": "email",
          "cell": [Function],
          "header": "email",
        },
        Object {
          "accessorKey": "phoneNumber",
          "cell": [Function],
          "header": "phone number",
        },
        Object {
          "accessorFn": [Function],
          "cell": [Function],
          "header": "date joined",
        },
        Object {
          "accessorFn": [Function],
          "cell": [Function],
          "header": "status",
        },
        Object {
          "accessorKey": "id",
          "cell": [Function],
          "header": [Function],
        },
      ]
    }
    data={null}
  />
</section>
`;
